kiosk/kiosk_browser/proxy.py

```diff
@@ -0,0 +1,124 @@
+
+    if 'Method' in config:
+        method = config['Method']
+        if method == 'direct':
+            return None
+        elif method == 'manual':
+            if 'Servers' in config:
+                servers = config['Servers']
+                if len(servers) >= 1:
+                    return parse_url(servers[0])
+    else:
+        return None
+
+def get_current_proxy(bus):
+    """Get current proxy from dbus
+
+    Return the proxy of a connected service preferentially, or of a ready
+    service.
+    """
+
+    client = dbus.Interface(
+        bus.get_object('net.connman', '/'),
+        'net.connman.Manager')
+
+    # List services, each service is a (id, properties) struct
+    services = client.GetServices()
+
+    # The service with the default route will always be sorted at the top of
+    # the list. (From connman doc/overview-api.txt)
+    default_service = find(
+        lambda s: s[1]['State'] == 'online' or s[1]['State'] == 'ready',
+        services)
+
+    if default_service:
+        return extract_manual_proxy(default_service[1]['Proxy'])
+
+def find(f, xs):
+    return next((x for x in xs if f(x)), None)
+
+def set_proxy_in_qt_app(hostname, port):
+    logging.info(f"Set proxy to {hostname}:{port} in Qt application")
+    network_proxy = QNetworkProxy()
+    network_proxy.setType(QNetworkProxy.HttpProxy)
+    network_proxy.setHostName(hostname)
+    network_proxy.setPort(port)
+    QNetworkProxy.setApplicationProxy(network_proxy)
+
+def set_no_proxy_in_qt_app():
+    logging.info(f"Set no proxy in Qt application")
+    QNetworkProxy.setApplicationProxy(QNetworkProxy())
+
+class Proxy():
+
+    def __init__(self):
+        DBusGMainLoop(set_as_default=True)
+        self._bus = dbus.SystemBus()
+        self._proxy = get_current_proxy(self._bus)
+
+    def start_monitoring_daemon(self):
+        """Use initial proxy in Qt application and watch for changes."""
+        self._use_in_qt_app()
+        thread = threading.Thread(target=self._monitor, args=[])
+        thread.daemon = True
+        thread.start()
+
+    def get_current(self):
+        return self._proxy
+
+    def _monitor(self):
+        self._bus.add_signal_receiver(
+            handler_function = self._on_property_changed,
+            bus_name = 'net.connman',
+            member_keyword = 'PropertyChanged')
+
+        # Update just after monitoring is on, so that we do not miss any proxy
+        # modification that could have happen before.
+        self._update(get_current_proxy(self._bus))
+
+        loop = GLib.MainLoop()
+        loop.run()
+
+    def _on_property_changed(self, *args, **kwargs):
+        if len(args) >= 2 and args[0] == 'Proxy':
+            self._update(extract_manual_proxy(args[1]))
```

Comment:
I think this causes a problem because it will update the proxy configuration to that of _any_ service for which a property changes.

This means, to give a very relevant example, that when connecting a Senso to the Play Computer after it has already started, the proxy configuration will be unset in the kiosk.

Maybe the following would be better?

```suggestion
            self._update(get_current_proxy(self._bus))
```